From 639aa47d17134ab658ae1be2a7ef24819ba06378 Mon Sep 17 00:00:00 2001
From: James McCoy <jamessan@jamessan.com>
Date: Sun, 30 Oct 2022 17:02:18 -0400
Subject: [PATCH] Update to notify v5 via notify_debouncer_mini

v5 notify moved the debounced API into its own crate,
notify_debouncer_mini.  The debounced API doesn't provide details on the
type of event that happened, just that a list of events or errors
happened.  Therefore, trigger reload on any event for a matching path.

Disable default features for notify_debouncer_mini since std::sync::mpsc
is being used, so there's no need to pull in crossbeam-channel.
---
 Cargo.lock                      | 78 ++++++++++++++++++++++++---------
 alacritty/Cargo.toml            |  2 +-
 alacritty/src/config/monitor.rs | 24 +++++-----
 3 files changed, 69 insertions(+), 35 deletions(-)

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -73,8 +73,9 @@ features = [
     "serde",
 ]
 
-[dependencies.notify]
-version = "4"
+[dependencies.notify-debouncer-mini]
+version = "0.2.1"
+default-features = false
 
 [dependencies.once_cell]
 version = "1.12"
--- a/src/config/monitor.rs
+++ b/src/config/monitor.rs
@@ -4,7 +4,8 @@ use std::time::Duration;
 
 use glutin::event_loop::EventLoopProxy;
 use log::{debug, error};
-use notify::{watcher, DebouncedEvent, RecursiveMode, Watcher};
+use notify_debouncer_mini::new_debouncer;
+use notify_debouncer_mini::notify::RecursiveMode;
 
 use alacritty_terminal::thread;
 
@@ -40,8 +41,8 @@ pub fn watch(mut paths: Vec<PathBuf>, ev
 
     // The Duration argument is a debouncing period.
     let (tx, rx) = mpsc::channel();
-    let mut watcher = match watcher(tx, DEBOUNCE_DELAY) {
-        Ok(watcher) => watcher,
+    let mut debouncer = match new_debouncer(DEBOUNCE_DELAY, None, tx) {
+        Ok(debouncer) => debouncer,
         Err(err) => {
             error!("Unable to watch config file: {}", err);
             return;
@@ -61,6 +62,7 @@ pub fn watch(mut paths: Vec<PathBuf>, ev
         parents.sort_unstable();
         parents.dedup();
 
+        let watcher = debouncer.watcher();
         // Watch all configuration file directories.
         for parent in &parents {
             if let Err(err) = watcher.watch(&parent, RecursiveMode::NonRecursive) {
@@ -69,26 +71,20 @@ pub fn watch(mut paths: Vec<PathBuf>, ev
         }
 
         loop {
-            let event = match rx.recv() {
-                Ok(event) => event,
+            let res = match rx.recv() {
+                Ok(res) => res,
                 Err(err) => {
                     debug!("Config watcher channel dropped unexpectedly: {}", err);
                     break;
                 },
             };
 
-            match event {
-                DebouncedEvent::Rename(_, path)
-                | DebouncedEvent::Write(path)
-                | DebouncedEvent::Create(path)
-                | DebouncedEvent::Chmod(path)
-                    if paths.contains(&path) =>
-                {
+            if let Ok(events) = res {
+                if events.iter().any(|e| paths.contains(&e.path)) {
                     // Always reload the primary configuration file.
                     let event = Event::new(EventType::ConfigReload(paths[0].clone()), None);
                     let _ = event_proxy.send_event(event);
-                },
-                _ => (),
+                }
             }
         }
     });
